### PR TITLE
Add case function for case-when type handling in math expression

### PR DESCRIPTION
**Pull request by navis**
support case-when-else clause in math expression

### CODE REVIEW

File: common/src/main/java/io/druid/math/expr/Function.java [hunk 1]
```diff
@@ -62,3 +62,24 @@ public Number apply(List<Expr> args, Map<String, Number> bindings)
...
+    for (int i = 0; i < args.size(); i += 2) {
+      Number x = args.get(i).eval(bindings);
+      if (x instanceof Long && x.longValue() > 0 || x.doubleValue() > 0) {
+        return args.get(i + 1).eval(bindings);
+      }
+    }
+    if (args.size() % 2 != 0) {
+      return args.get(args.size() - 1).eval(bindings);
+    }
+    return 0;
```
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/Function.java]: this is an implicit `ELSE 0`, right?

Comment (gianm) [inline common/src/main/java/io/druid/math/expr/Function.java]: perhaps include a test for this case.

File: common/src/main/antlr4/io/druid/math/expr/antlr/Expr.g4 [hunk 2]
```diff
@@ -36,3 +43,30 @@ EQ : '==' ;
 NEQ : '!=' ;
 AND : '&&' ;
 OR : '||' ;
+
+fragment A:('a'|'A');
```
Comment (gianm) [inline common/src/main/antlr4/io/druid/math/expr/antlr/Expr.g4]: haha, is this how you do case-insensitivity in antlr? :)

**Comment by gianm** [comment]
👍

**Comment by CHOIJAEHONG1** [comment]
LGTM

**Comment by himanshug** [comment]
@navis can this be done by a case function like the "if" function instead of updating the grammar? trying to keep the language simple.

File: common/src/main/java/io/druid/math/expr/Function.java (hunk 1 above)
Comment (navis) [inline common/src/main/java/io/druid/math/expr/Function.java]: sure.

**Comment by navis** [comment]
@himanshug I don't know well on antlr. Could you do that?

**Comment by himanshug** [comment]
@navis i meant the way "if" is implemented. it is a function only. so there could just be a function called "case" that accepts 1+2n+1 arguments. user would write something like
`case(case_expr, predicate1, then1, predicate2, then2, ..... else_expression)`

instead of 
`CASE case_expr WHEN predicate1 THEN then1 WHEN predicate2 THEN then2.... ELSE else_expression`

that would not be baking case syntax in the language grammar itself and possible keep parsing and language simple since this is supposed to be a tiny minimal language.

that said, i see this can be debated but just my 2 cents.

**Comment by navis** [comment]
@himanshug I prefer SQL like grammar but respect your intention. I'll do that.

**Comment by fjy** [comment]
@navis merge conflicts

**Pull request closed without merging**